Decay habitat quality threats using euclidean distance transform

dcdenu4 opened the pull request:
## Description
The Habitat Quality model has used convolutions as an implementation for decaying threat rasters over distance since InVEST 3.3.0. This approach strayed from the implementation described in the User's Guide and the User's Guide was never updated to reflect it. When thinking about updating the User's Guide to mention the convolution implementation it was not clear that the exponential decay via convolutions was providing the desired result. The degradation raster produced from the filtered threat rasters did not make intuitive sense either. Thus a decision was made to fall back to a euclidean distance transform implementation.

- Tests have been "eye" tested and updated.
- Model has been run using the Workbench UI and spot checked in GIS
- User's Guide has a PR here: https://github.com/natcap/invest.users-guide/pull/109

Fixes #1104 

## Checklist
- [x] Updated HISTORY.rst and link to any relevant issue (if these changes are user-facing)
- [x] Updated the user's guide (if needed)
- [x] Tested the affected models' UIs (if relevant)

phargogh commented on HISTORY.rst:
```diff
@@ -41,6 +41,17 @@ Unreleased Changes
+* Habitat Quality
+    * The model now uses an euclidean distance implementation for decaying
+      threat rasters both linearly and exponentially. Since InVEST 3.3.0 a
+      convolution implemenation has been used, which reflected how
```
```suggestion
      convolution implementation has been used, which reflected how
```

phargogh commented on tests/test_habitat_quality.py:
```diff
@@ -258,7 +258,7 @@ def test_habitat_quality_regression_different_projections(self):
         access_vector_path = os.path.join(
             args['workspace_dir'], 'access_samp.shp')
-        make_access_shp('access_vector_path')
+        make_access_shp(access_vector_path)
```
Was this always writing to CWD before??

dcdenu4 replied:
Haha, yes!!!

phargogh commented on src/natcap/invest/habitat_quality.py:
```diff
@@ -2,6 +2,7 @@
 """InVEST Habitat Quality model."""
 import collections
 import logging
+import math
```
Is `math` used?

phargogh replied:
`flake8` is also suggesting that `osr` is imported but unused.

dcdenu4 replied:
Thanks, I need to incorporate flake8 back into my workflow

dcdenu4 replied:
Fixed https://github.com/natcap/invest/pull/1159/commits/30ba06614bd613c6870a36a4b8b4a0562cf2454b

phargogh commented on src/natcap/invest/habitat_quality.py:
```diff
@@ -1013,6 +913,78 @@ def _raster_values_in_bounds(raster_path_band, lower_bound, upper_bound):
+def _decay_distance(dist_raster_path, max_dist, decay_type, target_path):
+    """Apply an exponential or linear decay to a distance transform raster.
+
+    The function will set pixles greater than ``max_dist`` to 0.
+
+    Args:
+        dist_raster_path (string): a filepath for the raster to decay.
+            The raster is expected to be a euclidean distance transform with
+            values meauring distance in pixels.
```
```suggestion
    The function will set pixels greater than ``max_dist`` to 0.

    Args:
        dist_raster_path (string): a filepath for the raster to decay.
            The raster is expected to be a euclidean distance transform with
            values measuring distance in pixels.
```

phargogh commented on src/natcap/invest/habitat_quality.py:
```diff
@@ -1013,6 +913,78 @@ def _raster_values_in_bounds(raster_path_band, lower_bound, upper_bound):
+
+    # convert max distance from meters to the number of pixels that
+    # represents on the raster
+    max_dist_pixel = max_dist_m / abs(threat_pixel_size[0])
```
Ah ok great, I see that these are already assumed to be square from the `align_and_resize_raster_stack` step.

phargogh commented on src/natcap/invest/habitat_quality.py:
```diff
@@ -1162,3 +1134,4 @@ def validate(args, limit_to=None):
                 invalid_keys.add('threats_table_path')
 
     return validation_warnings
+
```
Bonus newline?

dcdenu4 replied:
Fixed https://github.com/natcap/invest/pull/1159/commits/30ba06614bd613c6870a36a4b8b4a0562cf2454b

phargogh commented on tests/test_habitat_quality.py:
```diff
@@ -1048,9 +1048,12 @@ def test_habitat_quality_case_insensitivty(self):
         habitat_quality.execute(args)
 
         # Reasonable to just check quality out in this case
+        #assert_array_sum(
+        #    os.path.join(args['workspace_dir'], 'quality_c.tif'),
+        #    5852.088)
```
Was this left over from debugging, or should we still be asserting this?

dcdenu4 replied:
Huh, it must have been from debugging the new test results. It's a duplicate to an assertion right below it. I removed it. https://github.com/natcap/invest/pull/1159/commits/85a6d23583e268313d5838cea43b069a18882d60

phargogh reviewed (changes requested):
Looks sensible to me, @dcdenu4 !  I had a variety of trivial edits to suggest, and then it looks like the HQ tests are failing on something that looks like a missing raster.

Would you like to wait until https://github.com/natcap/invest.users-guide/pull/109 is merged and the UG revision here updated?

dcdenu4 commented:
Hey @phargogh, I fixed the breaking test in https://github.com/natcap/invest/pull/1159/commits/8d177219feb59926e176589ddbccf8d9b8f22c60. I think I addressed all your other comments as well.

I do think it'd be nice to wait for the UG PR to merge to be able to link this to that revision. I'm just waiting for Stacie to give me a thumbs up on a new figure.

dcdenu4 commented:
@phargogh, I've updated the Makefile with the UG revision corresponding to the merged UG changes!

dcdenu4 commented:
Just noting that we're merging this into `main` where as the UG revision is from UG release branch 3.13.0 where GLOBIO has been removed. GLOBIO is not removed in InVEST `main` yet though, it's sitting in the InVEST 3.13.0 branch. In case that breaks anything...

phargogh commented:
@dcdenu4 There's still one action failure, when building/validating sample data.  Other than that, I think this is good to go!

dcdenu4 commented:
@phargogh, that failure is from the GLOBIO mismatch between InVEST `main` (GLOBIO exists still) and UG `3.13.0` (GLOBIO is removed). I could switch the base in this PR to merge into InVEST `3.13.0` instead of `main`... I'll do that.

dcdenu4 commented:
Wait for PR #1206 before resolving conflicts and merging.

dcdenu4 commented:
Hey @phargogh , I _THINK_ that once the rest of the tests pass this should be good to go. 🤞
